In [188]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt 
import seaborn as sns   

In [189]:
df1 = pd.read_csv('../datos/procesados/df1.csv') 
df1.head()  

,ID_TRANSACCION,ID_CLIENTE,EDAD,SEXO,TIPO_CLIENTE,SEGMENTO,ANTIGUEDAD_MESES,INGRESOS_MENSUALES,EGRESOS_MENSUALES,MONTO_TRANSACCION,...,FRAUDE,HORA,RANGO_DISTANCIA_UBICACION_KM,RANGO_INGRESOS,RANGO_EGRESOS,RANGO_EDAD,RANGO_NUM_TRANSACCIONES_DIA,RANGO_MONTO_PROMEDIO_30D,RANGO_MONTO_TRANSACCION,RANGO_NUM_INTENTOS_FALLIDOS
0,TX0012629,CLI02084,60,M,Microempresario,Comercio,15,3913655.0,3148612.0,186052,...,0,18,20km - 50km,2M - 5M,2M - 5M,61 - 70,0 - 5,Hasta $250K,Hasta $250K,0-4000
1,TX0037731,CLI03274,54,F,Pequeño empresario,Microempresa,24,10757188.0,1613578.0,1200273,...,0,4,2km - 5km,10M - 20M,Hasta 2M,61 - 70,0 - 5,$1M - $2M,$1M - $2M,0-4000
2,TX0039992,CLI12186,44,M,Microempresario,Emprendimiento,23,2436868.0,1732651.0,242589,...,0,12,2km - 5km,2M - 5M,Hasta 2M,41 - 50,0 - 5,$250K - $500K,Hasta $250K,0-4000
3,TX0008526,CLI04248,22,F,Microempresario,Servicios,15,11709479.0,7468662.0,2490968,...,0,11,2km - 5km,10M - 20M,5M - 10M,21 - 30,0 - 5,$1M - $2M,$2M - $5M,0-4000
4,TX0008280,CLI11936,52,M,Pequeño empresario,Microempresa,10,10128545.0,6640160.0,334308,...,0,12,10km - 20km,10M - 20M,5M - 10M,61 - 70,0 - 5,$250K - $500K,$250K - $500K,0-4000


**Respuestas a preguntas de negocio**

1. ¿Cuál es la proporción de operaciones fraudulentas respecto al total?

In [190]:
proporcion_fraude = df1['FRAUDE'].mean() * 100    
print(f"Proporción de operaciones fraudulentas es de: {proporcion_fraude:.2f}%")

Proporción de operaciones fraudulentas es de: 4.28%


2. ¿Qué segmentos presentan mayor tasa de fraude?

In [191]:
tasa_segmento = (df1.groupby('SEGMENTO')['FRAUDE']
    .agg(['count', 'sum', 'mean'])
    .rename(columns={'count': 'Total_Operaciones', 'sum': 'Total_Fraudes',
        'mean': 'Tasa_Fraude'}))

tasa_segmento['Tasa_Fraude'] = (
    tasa_segmento['Tasa_Fraude'] * 100).round(2)

tasa_segmento = tasa_segmento.sort_values(
    'Tasa_Fraude', ascending=False)

tasa_segmento


,Total_Operaciones,Total_Fraudes,Tasa_Fraude
SEGMENTO,,,
Preferencial,4757,239,5.02
Comercio,9576,431,4.50
Microempresa,22770,964,4.23
Emprendimiento,13219,544,4.12
Servicios,9678,389,4.02


Los SEGMENTOS *Preferencial* con 5.02 y *Comercio* con 4.5, representan una tasa de fraude mayor a la proporcion de operaciones fraudulentas.  

3. ¿Qué canales presentan mayor incidencia de fraude?


In [192]:
tasa_canal = (df1.groupby('CANAL')['FRAUDE']
    .agg(['count', 'sum', 'mean'])
    .rename(columns={'count': 'Total_Operaciones', 'sum': 'Total_Fraudes',
        'mean': 'Tasa_Fraude'}))

tasa_canal['Tasa_Fraude'] = (
    tasa_canal['Tasa_Fraude'] * 100).round(2)

tasa_canal = tasa_canal.sort_values(
    'Tasa_Fraude', ascending=False)

tasa_canal

,Total_Operaciones,Total_Fraudes,Tasa_Fraude
CANAL,,,
App,24991,1254,5.02
Web,13258,519,3.91
Oficina,8002,308,3.85
Cajero,4819,177,3.67
Corresponsal,8930,309,3.46


El CANAL que presenta mayor incidencia de fraude es *App* con 5.02, que es mayor a la proprocion de operaciones fraudulentas 

4. ¿Qué tipos de transacción presentan mayor riesgo?

In [193]:
tasa_tipo_transacción = (df1.groupby('TIPO_TRANSACCION')['FRAUDE']
    .agg(['count', 'sum', 'mean'])
    .rename(columns={'count': 'Total_Operaciones', 'sum': 'Total_Fraudes',
        'mean': 'Tasa_Fraude'}))

tasa_tipo_transacción['Tasa_Fraude'] = (
    tasa_tipo_transacción['Tasa_Fraude'] * 100).round(2)

tasa_tipo_transacción = tasa_tipo_transacción.sort_values(
    'Tasa_Fraude', ascending=False)

tasa_tipo_transacción

,Total_Operaciones,Total_Fraudes,Tasa_Fraude
TIPO_TRANSACCION,,,
Transferencia,16701,848,5.08
Desembolso,6008,256,4.26
Retiro,7302,302,4.14
Compra,9532,382,4.01
Pago proveedor,10929,416,3.81
Pago factura,9528,363,3.81


El TIPO_TRANSACCION que representa mayor incidencia de fraude es *Transferencia* con 5.08, que es mayor a la proprocion de operaciones fraudulentas y luego esta *Desembolso* con 4.26.

5. ¿Existen diferencias de fraude entre ciudades?

In [194]:
fraude_ciudad = (df1.groupby('CIUDAD')['FRAUDE'].agg(Total_Operaciones='count',
        Total_Fraudes='sum', Tasa_Fraude=lambda x: x.mean() * 100)
    .sort_values('Tasa_Fraude', ascending=False))

fraude_ciudad['Tasa_Fraude'] = fraude_ciudad['Tasa_Fraude'].round(2)

fraude_ciudad

,Total_Operaciones,Total_Fraudes,Tasa_Fraude
CIUDAD,,,
Bello,5964,302,5.06
Barranquilla,4123,195,4.73
Bogotá,12037,512,4.25
Medellín,16808,714,4.25
Cali,5941,252,4.24
Manizales,3038,127,4.18
Rionegro,2903,119,4.10
Bucaramanga,3005,118,3.93
Pereira,3126,116,3.71


Los datos revelan que Bello con 5.06 y Barranquilla con 4.73 representan una tasa de fraude superior a la proporcion de operaciones fraudulentas 

6. ¿Existen diferencias según país de destino?

In [195]:
fraude_pais_destino = (df1.groupby('PAIS_DESTINO')['FRAUDE'].agg(Total_Operaciones='count',
        Total_Fraudes='sum', Tasa_Fraude=lambda x: x.mean() * 100)
    .sort_values('Tasa_Fraude', ascending=False))

fraude_pais_destino['Tasa_Fraude'] = fraude_pais_destino['Tasa_Fraude'].round(2)

fraude_pais_destino

,Total_Operaciones,Total_Fraudes,Tasa_Fraude
PAIS_DESTINO,,,
MX,615,46,7.48
US,2092,143,6.84
EC,1234,60,4.86
CO,54606,2263,4.14
PA,1453,55,3.79


Para la cantidad de operaciones fraudulentas que maneja cada pais, MX con 7,48, US con 6.84 y EC con 4.86 en su tasa de fraude, se observa que estos representan una tasa mayor de fraude respecto a la proporcion de operaciones fraudulentas, incluso asi tengan un menor numero de fraudes.

7. ¿Qué rangos de edad presentan mayor tasa de fraude?

In [196]:
tasa_rango_edad = (df1.groupby('RANGO_EDAD')['FRAUDE']
    .agg(['count', 'sum', 'mean'])
    .rename(columns={'count': 'Total_Operaciones', 'sum': 'Total_Fraudes',
        'mean': 'Tasa_Fraude'}))

tasa_rango_edad['Tasa_Fraude'] = (
    tasa_rango_edad['Tasa_Fraude'] * 100).round(2)

tasa_rango_edad = tasa_rango_edad.sort_values(
    'Tasa_Fraude', ascending=False)

tasa_rango_edad

,Total_Operaciones,Total_Fraudes,Tasa_Fraude
RANGO_EDAD,,,
Hasta 20,2811,133,4.73
21 - 30,9491,443,4.67
41 - 50,17875,812,4.54
Más de 70,246,11,4.47
61 - 70,10510,441,4.20
31 - 40,19067,727,3.81


Los datos muestran que hay un rango amplio de edades que sobrepasan la proporcion de fraude total que esta en 4.28. Por lo tanto la tasa de fraude para las edades entre los 31 - 40 con 3.81 y entre los 61 - 70 años con 4.20 estan dentro del rango normal.

8. ¿Qué rangos de ingresos presentan mayor tasa de fraude?

In [197]:
tasa_rango_ingresos = (df1.groupby('RANGO_INGRESOS')['FRAUDE']
    .agg(['count', 'sum', 'mean'])
    .rename(columns={'count': 'Total_Operaciones', 'sum': 'Total_Fraudes',
        'mean': 'Tasa_Fraude'}))

tasa_rango_ingresos['Tasa_Fraude'] = (
    tasa_rango_ingresos['Tasa_Fraude'] * 100).round(2)

tasa_rango_ingresos = tasa_rango_ingresos.sort_values(
    'Tasa_Fraude', ascending=False)

tasa_rango_ingresos

,Total_Operaciones,Total_Fraudes,Tasa_Fraude
RANGO_INGRESOS,,,
Más de 50M,286,19,6.64
20M - 50M,3533,157,4.44
10M - 20M,10575,468,4.43
5M - 10M,18237,778,4.27
2M - 5M,20278,856,4.22
Hasta 2M,7091,289,4.08


La tasa de fraude para el rango de ingresos > a 50 millones esta significativamente alta de acuerdo con la cantidad total de operaciones. 6.64 resepcto a los 4.28 del totla de operaciones fraudulentas. 

9. ¿Existe relación entre el monto de la operación y el fraude?

In [198]:
fraude_monto_transacción = (df1.groupby('RANGO_MONTO_TRANSACCION')['FRAUDE'].agg(Total_Operaciones='count',
        Total_Fraudes='sum', Tasa_Fraude=lambda x: x.mean())
    .sort_values('Tasa_Fraude', ascending=False))

fraude_monto_transacción['Tasa_Fraude'] = (
    fraude_monto_transacción['Tasa_Fraude'] * 100
 ).round(2)

fraude_monto_transacción

,Total_Operaciones,Total_Fraudes,Tasa_Fraude
RANGO_MONTO_TRANSACCION,,,
Más de $5M,1925,105,5.45
$2M - $5M,7205,357,4.95
$1M - $2M,10943,488,4.46
Hasta $250K,13724,580,4.23
$250K - $500K,12211,495,4.05
$500K - $1M,13992,542,3.87


Los datos indican que para montos de transacción superiores a 5 millones la tasa de fraude esta alta 5.45, seguida de montos entre 2 y 5 millones con un valor de 4.95

10. ¿Qué relación existe entre ingresos, egresos y monto transaccional?

In [199]:
corr_ing_egr_monto = df1[['INGRESOS_MENSUALES', 'EGRESOS_MENSUALES', 'MONTO_TRANSACCION']].corr()
corr_ing_egr_monto

,INGRESOS_MENSUALES,EGRESOS_MENSUALES,MONTO_TRANSACCION
INGRESOS_MENSUALES,1.000000,0.907632,0.630777
EGRESOS_MENSUALES,0.907632,1.000000,0.574616
MONTO_TRANSACCION,0.630777,0.574616,1.000000


Los datos muestran una alta correlacion entre Ingresos y egresos mensuales = 0.907 y una correlación de ambas con el monto de la transacción muy similar de 0.630 y 0.574 respectivamente.

11. ¿Las operaciones desde dispositivos nuevos presentan mayor riesgo?

In [200]:
riesgo_dispositivo = (df1.groupby('DISPOSITIVO_NUEVO')['FRAUDE'].agg(Total_Operaciones='count',
        Total_Fraudes='sum', Tasa_Fraude=lambda x: x.mean() * 100))

riesgo_dispositivo['Tasa_Fraude'] = (
    riesgo_dispositivo['Tasa_Fraude'].round(2))

riesgo_dispositivo

,Total_Operaciones,Total_Fraudes,Tasa_Fraude
DISPOSITIVO_NUEVO,,,
0,52822,1903,3.60
1,7178,664,9.25


Los datos indican que la tasa de fraude para esta categoria representa mas del doble del valor de la tasa de operaciones fraudulentas del total que es 4.28

12. ¿Los cambios recientes de clave están asociados al fraude?

In [201]:
cambio_clave = (df1.groupby('CAMBIO_CLAVE_RECIENTE')['FRAUDE'].agg(Total_Operaciones='count',
        Total_Fraudes='sum', Tasa_Fraude=lambda x: x.mean() * 100))

cambio_clave['Tasa_Fraude'] = (
    cambio_clave['Tasa_Fraude'].round(2))

cambio_clave

,Total_Operaciones,Total_Fraudes,Tasa_Fraude
CAMBIO_CLAVE_RECIENTE,,,
0,55289,2216,4.01
1,4711,351,7.45


Se encuentra que esta variable tambien representa una tasa de fraude alta, con un valor de 7.45

13. ¿La distancia geográfica está relacionada con operaciones fraudulentas?

In [202]:
distancia_ubicacion = (df1.groupby('RANGO_DISTANCIA_UBICACION_KM')['FRAUDE'].agg(Total_Operaciones='count',
        Total_Fraudes='sum', Tasa_Fraude=lambda x: x.mean() * 100))

distancia_ubicacion['Tasa_Fraude'] = (
    distancia_ubicacion['Tasa_Fraude'].round(2))

distancia_ubicacion

,Total_Operaciones,Total_Fraudes,Tasa_Fraude
RANGO_DISTANCIA_UBICACION_KM,,,
0km - 2km,3916,160,4.09
10km - 20km,23962,1055,4.40
20km - 50km,7357,325,4.42
2km - 5km,7853,302,3.85
5km - 10km,14644,599,4.09
Hasta 0km,1611,75,4.66
Más de 50km,657,51,7.76


La tasa de fraude en operaciones a más de 50 km es aproximadamente 1,8 veces la tasa global del dataset.

14. ¿Existe relación entre intentos fallidos y fraude?

In [203]:
intentos_fallidos = (df1.groupby('NUM_INTENTOS_FALLIDOS')['FRAUDE'].agg(Total_Operaciones='count',
        Total_Fraudes='sum', Tasa_Fraude=lambda x: x.mean() * 100))

intentos_fallidos['Tasa_Fraude'] = (
    intentos_fallidos['Tasa_Fraude'].round(2))

intentos_fallidos

,Total_Operaciones,Total_Fraudes,Tasa_Fraude
NUM_INTENTOS_FALLIDOS,,,
0,27043,892,3.30
1,21437,945,4.41
2,8685,495,5.70
3,2284,178,7.79
4,466,45,9.66
5,79,9,11.39
6,6,3,50.00


Sí existe una relación evidente entre el número de intentos fallidos y la tasa de fraude. La tasa de fraude aumenta progresivamente desde 3,30% en operaciones sin intentos fallidos hasta 11,39% en operaciones con cinco intentos fallidos. El valor de 50% observado para seis intentos debe interpretarse con precaución debido al reducido número de observaciones.

15. ¿Los clientes nuevos presentan mayor incidencia?

In [204]:
tasa_beneficiario_nuevo = (df1.groupby('CLIENTE_NUEVO')['FRAUDE'].agg(Total_Operaciones='count',
        Total_Fraudes='sum', Tasa_Fraude=lambda x: x.mean() * 100))

tasa_beneficiario_nuevo['Tasa_Fraude'] = (
    tasa_beneficiario_nuevo['Tasa_Fraude'].round(2))

tasa_beneficiario_nuevo

,Total_Operaciones,Total_Fraudes,Tasa_Fraude
CLIENTE_NUEVO,,,
0,58789,2489,4.23
1,1211,78,6.44


Los clientes nuevos presentan incidencia de fraude en un 6.44 frente al 4.23 de los clientes de mayor antiguedad


16. ¿Las operaciones nocturnas presentan mayor riesgo?

In [205]:
df1['OPERACION_NOCTURNA'] = ((df1['HORA'] >= 18) | (df1['HORA'] < 6)).astype(int)


tasa_operación_noct = (df1.groupby('OPERACION_NOCTURNA')['FRAUDE'].agg(Total_Operaciones='count',
        Total_Fraudes='sum', Tasa_Fraude=lambda x: x.mean() * 100))

tasa_operación_noct['Tasa_Fraude'] = (
    tasa_operación_noct['Tasa_Fraude'].round(2))

tasa_operación_noct

,Total_Operaciones,Total_Fraudes,Tasa_Fraude
OPERACION_NOCTURNA,,,
0,30006,1097,3.66
1,29994,1470,4.90


Los datos indican que las operaciones nocturnas persentan una mayor tasa de fraude con un 4.9.

17. ¿Los beneficiarios nuevos presentan mayor incidencia de fraude?

In [206]:
tasa_beneficiario_nuevo = (df1.groupby('BENEFICIARIO_NUEVO')['FRAUDE'].agg(Total_Operaciones='count',
        Total_Fraudes='sum', Tasa_Fraude=lambda x: x.mean() * 100))

tasa_beneficiario_nuevo['Tasa_Fraude'] = (
    tasa_beneficiario_nuevo['Tasa_Fraude'].round(2))

tasa_beneficiario_nuevo

,Total_Operaciones,Total_Fraudes,Tasa_Fraude
BENEFICIARIO_NUEVO,,,
0,51569,1953,3.79
1,8431,614,7.28


Los beneficiarios nuevos represantan una tasa de fraude muy alta, 7.28 para un margen de operaciones tan bajo. 

18. ¿Qué variables parecen tener mayor poder discriminativo?

El análisis exploratorio identifica como principales variables candidatas para discriminar entre operaciones fraudulentas y no fraudulentas NUM_INTENTOS_FALLIDOS, BENEFICIARIO_NUEVO, DISTANCIA_UBICACION_KM, CLIENTE_NUEVO y OPERACION_NOCTURNA. Entre ellas, el número de intentos fallidos presenta uno de los patrones más claros, mostrando un incremento progresivo de la tasa de fraude a medida que aumenta el número de intentos. Estos resultados son exploratorios y deberán validarse posteriormente mediante técnicas de selección de variables e importancia predictiva del modelo.

**Variables que quedan para el modelo**

In [211]:
from pathlib import Path
import sys

root_dir = Path.cwd().resolve().parent if Path.cwd().name == 'notebooks' else Path.cwd().resolve()
if str(root_dir) not in sys.path:
    sys.path.append(str(root_dir))

from funciones.variables_modelo import (
    VARIABLES_CATEGORICAS_MODELO,
    VARIABLES_NUMERICAS,
    VARIABLES_TOTALES_MODELO,
 )

variables_numericas = VARIABLES_NUMERICAS.copy()
variables_categoricas = VARIABLES_CATEGORICAS_MODELO.copy()
variables_totales = VARIABLES_TOTALES_MODELO.copy()

Para seleccionar las variables con mayor poder discriminativo se combinan tres criterios: Cramér's V para variables categóricas, correlación de Pearson (en valor absoluto) para variables numéricas, y Mutual Information, que aplica por igual a ambos tipos de variable frente a `FRAUDE`. Cada variable se ordena según su métrica nativa y según su Mutual Information; el promedio de ambos rankings define un score combinado que permite comparar variables numéricas y categóricas en una misma escala.

In [212]:
from scipy.stats import chi2_contingency

def cramers_v(x, y):
    tabla = pd.crosstab(x, y)
    chi2 = chi2_contingency(tabla)[0]
    n = tabla.sum().sum()
    phi2 = chi2 / n
    r, k = tabla.shape
    phi2corr = max(0, phi2 - ((k - 1) * (r - 1)) / (n - 1))
    rcorr = r - ((r - 1) ** 2) / (n - 1)
    kcorr = k - ((k - 1) ** 2) / (n - 1)
    return np.sqrt(phi2corr / min((kcorr - 1), (rcorr - 1)))

cramers_scores = {col: cramers_v(df1[col], df1['FRAUDE']) for col in variables_categoricas}
cramers_scores = pd.Series(cramers_scores, name='Cramers_V').sort_values(ascending=False)
cramers_scores

DISPOSITIVO_NUEVO        0.090354
BENEFICIARIO_NUEVO       0.059770
CAMBIO_CLAVE_RECIENTE    0.045423
CANAL                    0.030594
OPERACION_NOCTURNA       0.030407
PAIS_DESTINO             0.028602
TIPO_TRANSACCION         0.023865
CLIENTE_NUEVO            0.014481
CIUDAD                   0.011950
SEGMENTO                 0.010143
FIN_DE_SEMANA            0.001566
SEXO                     0.000000
TIPO_CLIENTE             0.000000
Name: Cramers_V, dtype: float64

In [ ]:
correlacion_scores = (
    df1[variables_numericas + ['FRAUDE']]
    .corr()['FRAUDE']
    .drop('FRAUDE')
    .abs()
    .sort_values(ascending=False)
)
correlacion_scores.name = 'Correlacion_Abs'
correlacion_scores

NUM_INTENTOS_FALLIDOS     0.060772
MONTO_TRANSACCION         0.014065
DISTANCIA_UBICACION_KM    0.013922
ANTIGUEDAD_MESES          0.007629
INGRESOS_MENSUALES        0.007019
EGRESOS_MENSUALES         0.006651
MONTO_PROMEDIO_30D        0.006265
EDAD                      0.001536
NUM_TRANSACCIONES_DIA     0.000786
Name: Correlacion_Abs, dtype: float64

In [ ]:
from sklearn.feature_selection import mutual_info_classif
from sklearn.preprocessing import LabelEncoder

X_mi = df1[variables_totales].copy()
for col in variables_categoricas:
    X_mi[col] = LabelEncoder().fit_transform(X_mi[col].astype(str))

es_discreta = [col in variables_categoricas for col in variables_totales]

mi_scores = mutual_info_classif(
    X_mi, df1['FRAUDE'], discrete_features=es_discreta, random_state=42
)
mi_scores = pd.Series(mi_scores, index=variables_totales, name='Mutual_Info').sort_values(ascending=False)
mi_scores

DISPOSITIVO_NUEVO         0.003260
NUM_INTENTOS_FALLIDOS     0.002852
NUM_TRANSACCIONES_DIA     0.002339
MONTO_TRANSACCION         0.001728
BENEFICIARIO_NUEVO        0.001552
CAMBIO_CLAVE_RECIENTE     0.000880
DISTANCIA_UBICACION_KM    0.000798
INGRESOS_MENSUALES        0.000720
CANAL                     0.000499
OPERACION_NOCTURNA        0.000475
ANTIGUEDAD_MESES          0.000461
PAIS_DESTINO              0.000379
TIPO_TRANSACCION          0.000318
CIUDAD                    0.000144
CLIENTE_NUEVO             0.000103
SEGMENTO                  0.000082
EGRESOS_MENSUALES         0.000026
FIN_DE_SEMANA             0.000010
TIPO_CLIENTE              0.000009
SEXO                      0.000003
EDAD                      0.000000
MONTO_PROMEDIO_30D        0.000000
Name: Mutual_Info, dtype: float64

In [ ]:
tabla_discriminativa = pd.DataFrame(index=variables_totales)
tabla_discriminativa['Tipo'] = ['Numerica' if v in variables_numericas else 'Categorica' for v in variables_totales]
tabla_discriminativa['Cramers_V'] = cramers_scores
tabla_discriminativa['Correlacion_Abs'] = correlacion_scores
tabla_discriminativa['Mutual_Info'] = mi_scores

tabla_discriminativa['Metrica_Nativa'] = tabla_discriminativa['Cramers_V'].fillna(tabla_discriminativa['Correlacion_Abs'])
tabla_discriminativa['Rank_MI'] = tabla_discriminativa['Mutual_Info'].rank(ascending=False)
tabla_discriminativa['Rank_Nativo'] = tabla_discriminativa['Metrica_Nativa'].rank(ascending=False)
tabla_discriminativa['Score_Combinado'] = tabla_discriminativa[['Rank_MI', 'Rank_Nativo']].mean(axis=1)

tabla_discriminativa = tabla_discriminativa.sort_values('Score_Combinado')
tabla_discriminativa

,Tipo,Cramers_V,Correlacion_Abs,Mutual_Info,Metrica_Nativa,Rank_MI,Rank_Nativo,Score_Combinado
DISPOSITIVO_NUEVO,Categorica,0.090354,NaN,0.003260,0.090354,1.0,1.0,1.00
NUM_INTENTOS_FALLIDOS,Numerica,NaN,0.060772,0.002852,0.060772,2.0,2.0,2.00
BENEFICIARIO_NUEVO,Categorica,0.059770,NaN,0.001552,0.059770,5.0,3.0,4.00
CAMBIO_CLAVE_RECIENTE,Categorica,0.045423,NaN,0.000880,0.045423,6.0,4.0,5.00
CANAL,Categorica,0.030594,NaN,0.000499,0.030594,9.0,5.0,7.00
MONTO_TRANSACCION,Numerica,NaN,0.014065,0.001728,0.014065,4.0,10.0,7.00
OPERACION_NOCTURNA,Categorica,0.030407,NaN,0.000475,0.030407,10.0,6.0,8.00
DISTANCIA_UBICACION_KM,Numerica,NaN,0.013922,0.000798,0.013922,7.0,11.0,9.00
PAIS_DESTINO,Categorica,0.028602,NaN,0.000379,0.028602,12.0,7.0,9.50
TIPO_TRANSACCION,Categorica,0.023865,NaN,0.000318,0.023865,13.0,8.0,10.50


In [ ]:
N_VARIABLES = 10
variables_seleccionadas = tabla_discriminativa.head(N_VARIABLES).index.tolist()

df_variables_discriminativas = df1[variables_seleccionadas + ['FRAUDE']].copy()
df_variables_discriminativas.to_csv('../datos/procesados/dataset_variables_discriminativas.csv', index=False)

print(f"Variables seleccionadas ({N_VARIABLES}): {variables_seleccionadas}")
df_variables_discriminativas.head()

Variables seleccionadas (10): ['DISPOSITIVO_NUEVO', 'NUM_INTENTOS_FALLIDOS', 'BENEFICIARIO_NUEVO', 'CAMBIO_CLAVE_RECIENTE', 'CANAL', 'MONTO_TRANSACCION', 'OPERACION_NOCTURNA', 'DISTANCIA_UBICACION_KM', 'PAIS_DESTINO', 'TIPO_TRANSACCION']


,DISPOSITIVO_NUEVO,NUM_INTENTOS_FALLIDOS,BENEFICIARIO_NUEVO,CAMBIO_CLAVE_RECIENTE,CANAL,MONTO_TRANSACCION,OPERACION_NOCTURNA,DISTANCIA_UBICACION_KM,PAIS_DESTINO,TIPO_TRANSACCION,FRAUDE
0,0,0,0,0,App,186052,1,81.92,CO,Pago proveedor,0
1,0,2,0,0,App,1200273,1,5.35,CO,Desembolso,0
2,0,0,0,0,Cajero,242589,0,9.13,CO,Pago factura,0
3,1,0,0,0,Cajero,2490968,0,7.69,CO,Transferencia,0
4,0,0,0,0,Cajero,334308,0,33.22,CO,Transferencia,0


El score combinado (promedio del ranking por Mutual Information y del ranking por la métrica nativa de cada tipo de variable) coincide con lo observado en el análisis exploratorio: las variables más discriminativas son señales de comportamiento reciente y de seguridad (`DISPOSITIVO_NUEVO`, `BENEFICIARIO_NUEVO`, `CAMBIO_CLAVE_RECIENTE`), el número de intentos fallidos (`NUM_INTENTOS_FALLIDOS`), y el contexto de la operación (`CANAL`, `OPERACION_NOCTURNA`, `PAIS_DESTINO`, `TIPO_TRANSACCION`), junto con `MONTO_TRANSACCION` y `DISTANCIA_UBICACION_KM` como variables numéricas relevantes. Variables demográficas como `SEXO`, `TIPO_CLIENTE` y `EDAD` quedan en las últimas posiciones, confirmando su bajo poder predictivo.

Con las 10 variables de mejor score combinado se generó el dataset `dataset_variables_discriminativas.csv` en `datos/procesados/`, que incluye `FRAUDE` como variable objetivo y queda listo para las siguientes fases de preparación y modelado.